index.js

```diff
@@ -8,6 +8,29 @@ function assertEventName(eventName) {
 	}
 }
 
+async function * iterator(emitter, eventName) {
+	const queue = [];
+	const off = emitter.on(eventName, data => {
+		queue.push(data);
+	});
+
+	try {
+		/* eslint-disable no-constant-condition */
+		/* eslint-disable no-await-in-loop */
+		while (true) {
+			if (queue.length > 0) {
+				yield queue.shift();
+			} else {
+				yield await emitter.once(eventName);
```

Comment:
This shouldn't yield, `queue` will contain the data of the next event after `await` returns.

Alternatively you could await a new promise, and expose its `resolve()` method to the listener created above. That way you don't need to add an additional listener.

Reply:
you are right the yield is superfluous. 
However I don't think exposing the resolve of the new promise on each iteration is a good idea: As the listener above is created once I would need to create a binding in its closure and reassign this binding to the resolve of the new promise on each iteration. That seems a bit awkward and unsafe, or I am missing something.